taichi/program/sparse_solver.cpp

```diff
@@ -32,17 +33,44 @@ bool EigenSparseSolver<EigenSolver>::info() {
   return solver_.info() == Eigen::Success;
 }
 
-std::unique_ptr<SparseSolver> get_sparse_solver(
-    const std::string &solver_type) {
-  if (solver_type == "LU") {
+#define MAKE_SOLVER(type, order)                                              \
+  {                                                                           \
+    {#type, #order}, []() -> std::unique_ptr<SparseSolver> {                  \
+      using T =                                                               \
+          Eigen::Simplicial##type<Eigen::SparseMatrix<float32>, Eigen::Lower, \
+                                  Eigen::order##Ordering<int>>;               \
+      return std::make_unique<EigenSparseSolver<T>>();                        \
+    }                                                                         \
+  }
+
+struct pair_hash {
```

Comment:
Final nit: place this at the top and guard it with an anonymous namespace (e.g. `namespace { ... } // namespace`)